### PR TITLE
Picking changes from the fix for SPARK-32709

### PR DESCRIPTION
**Pull request by s-pedamallu**
Applied patch from PR: https://github.com/apache/spark/pull/33432

Query in Hive: https://sql.lyft.net/databases/mozart/queries/92938921
Query in Spark: https://sql.lyft.net/databases/mozart/queries/92939005
Comparision: https://sql.lyft.net/databases/mozart/queries/92948392

However, the number of files being created in Spark is very large and probably the solution wouldn't work without custom configuration. Raised a request regarding the same in open-source: https://issues.apache.org/jira/browse/SPARK-32709?focusedCommentId=17396330&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17396330

**Context**
Spark previously did not support bucketed tables from Hive because the hashing function in hive is different from that of Spark. However, this issue was addressed by introducing hive hash in Spark and using it when writing data and hence this data written by Spark can now be efficiently be read by Hive and PrestoSql/Trino.

**TODO**
Although reducers use hive hash when writing output, the reducers are not partitioned using hive hash which is leading to many output files that could cause S3 throttling failures. Custom configuration by enabling adaptive query execution may help in some cases.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala [hunk 1]
```diff
@@ -222,17 +224,23 @@ class DynamicPartitionDataWriter(
...
 
     val customPath = partDir.flatMap { dir =>
       description.customPartitionLocations.get(PartitioningUtils.parsePathFragment(dir))
     }
     val currentPath = if (customPath.isDefined) {
-      committer.newTaskTempFileAbsPath(taskAttemptContext, customPath.get, ext)
+      committer.newTaskTempFileAbsPath(taskAttemptContext, customPath.get, fileNameSpec)
     } else {
-      committer.newTaskTempFile(taskAttemptContext, partDir, ext)
+      committer.newTaskTempFile(taskAttemptContext, partDir, fileNameSpec)
```
Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala:243]: Sorry to chime in here, but this PR depends on https://github.com/apache/spark/pull/33012. Shall we merge https://github.com/apache/spark/pull/33012 first? Thanks.

Comment (s-pedamallu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala:243]: apologies for late reply. but this change is already included as part of this PR.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala:243]: Yeah, my comment was published before your 2nd commit, so it looks good for me now.

**Comment by s-pedamallu** [comment]
👀 @catalinii @nanassito #compute-data-platform

**Review by catalinii — APPROVED** [review]

**Pull request merged**